### PR TITLE
Refactor request data clumping

### PR DESCRIPTION
**Pull request by will-in-wi**
Changes:
- Restore deprecated get/post/put/delete methods auto-prepending `v2/` to URL.
- Create `Request` object and refactor code to use it.
- Create `HTTPClient` object and refactor code to use it. Migrate relevant tests from `Client` to `HTTPClient`. This reduces the responsibilities of the main `Client` class dramatically.
- Move and make private `Drip::Client::REDIRECT_LIMIT` constant.

### CODE REVIEW

File: lib/drip/client.rb [hunk 1]
```diff
@@ -68,69 +68,31 @@ def content_type
...
-          end
-
-          response = http.request request
-          if response.is_a?(Net::HTTPRedirection)
-            return make_request(verb_klass, URI(response["Location"]), options, step + 1)
-          else
-            response
-          end
-        end
+        Drip::Client::HTTPClient.new(@config).make_request(drip_request)
```
Comment (rzaharenkov) [inline lib/drip/client.rb]: I'd suggest to create an instance of `Drip::Client::HTTPClient` once, memoize it and reset if config is changed.

File: lib/drip/client.rb [hunk 2]
```diff
@@ -68,69 +68,31 @@ def content_type
...
     end
 
-    def put(url, options = {})
-      make_request(Net::HTTP::Put, make_uri(url), options)
-    end
+  private
 
-    def delete(url, options = {})
-      make_request(Net::HTTP::Delete, make_uri(url), options)
+    def make_v2_request(http_verb, path, options = {})
```
Comment (rzaharenkov) [inline lib/drip/client.rb]: I think the name of the method doesn't reflects what it does. I'd suggest to have more generic `make_request` method that accepts different arguments, e.g:

```
def make_request(*args) ...
```

or

```
def make_request(request: nil, path: nil, verb: nil, options: nil) ...
```

File: lib/drip/client.rb (hunk 1 above)
Comment (will-in-wi) [reply lib/drip/client.rb]: What's the benefit of doing so? If we're going to be introducing some cache invalidation issues, I'd like to have a good reason.

File: lib/drip/client.rb (hunk 2 above)
Comment (will-in-wi) [reply lib/drip/client.rb]: Heh, that was the original version before my changes…

I'm struggling with what to do with the content_type, which is effectively the only difference. I played with a couple of possibilities:
1. Pass the content type with every invocation of make_request. This felt very repetitive.
2. Create `JSON` and `JSONAPI` subclasses of `Drip::Request` that hard-code the content type. This felt cleaner, but I didn't like making subclasses to encode data.
3. Create a builder method for each type. This is what I ultimately landed on, as it felt like a variant of option 1 with some syntactic sugar to make it cleaner.

It sounds like you are suggesting that I go back to option 1. Do you think that is cleaner? Is there another way?

Comment (rzaharenkov) [reply lib/drip/client.rb]: I think `make_json_request` makes a lot of difference from `make_v2_request`, I'd like such name much more.

File: lib/drip/client.rb (hunk 1 above)
Comment (rzaharenkov) [reply lib/drip/client.rb]: The point is you don't need to create a new instance of http client for each request. It leads to memory leaks and creates performance overhead. This object depends only on `@config` so I think it should be quite easy to reset it when config is changed.

File: lib/drip/client.rb (hunk 2 above)
Comment (will-in-wi) [reply lib/drip/client.rb]: I like that too. Change made!

**Pull request merged**